### PR TITLE
Cache normalized queries over SQL adapter

### PR DESCRIPTION
**Pull request by aljazerzen**
Enables query normalization for sql-over-pg. This uncovered a bunch of bugs that were also present on already enabled sql-over-edgedb-proto, but have not been caught because we didn't have many tests.

This PR also contains two commits to libpg_query. I suggest reviewing commit by commit.

Encountered problems:
- `select 1 + '1'` should infer that `'1'` has `integer` type. The solution was that string constants need to be reported to postgres as having type `unknown`.
- `SET TRANSACTION ISOLATION LEVEL read uncommitted` was being normalized to `SET TRANSACTION ISOLATION LEVEL $1`, which does not even parse. Solution was to patch libpg_query to only normalize SELECT, UPDATE, DELETE, INSERT and MERGE.
- `SELECT NULL, 42` was being normalized to `SELECT NULL, $2`. A bug in libpg_query (that we introduced, not the original author),
- `SELECT current_schemas(true)` was normalized to `SELECT current_schemas($1)`, which meant that we are unable to statically evaluate the query. Solution was `DisableNormalization` exception that re-compiles the query without normalization.
- `SELECT pg_get_serial_sequence('a', 'b')` was normalized to `SELECT pg_get_serial_sequence($1, $2)` and resolved to `SELECT NULL`, which tripped PostgreSQL, because it was unable to determine type of the two params, which we specified as `unknown`. The solution was to detect that the params are unused and report them as being `text`.
- `SELECT ROW('hello')` is normalized to `SELECT ROW($1)`, which tripped PostgreSQL, because it was unable to determine type of the `unknown` param. Solution was to compile fields in `ROW` with additional type cast to `text`.
- `SELECT to_json('hello')` is normalized to `SELECT to_json($1)`. Similar as above.

I still need to debug why all error spans have gone missing. I'd expect that they would just be wrong.

### CODE REVIEW

File: edb/server/compiler/sql.py [hunk 1]
```diff
@@ -89,15 +98,31 @@ def _try(q: str) -> List[dbstate.SQLQueryUnit]:
...
     try:
-        return _try(source.text())
+        try:
+            return _try(source.text(), normalized_params)
+        except DisableNormalization:
+            # compiler requested non-normalized query (it needs it for static
+            # evaluation)
+            try:
+                if isinstance(source, pg_parser.NormalizedSource):
+                    return _try(source.original_text(), [])
```
[the remark this answers is not in the document]
Comment (msullivan) [reply edb/server/compiler/sql.py:115]: Yeah, we do. I'll push a fix.

**Review by msullivan — APPROVED** [review]
Looks good. I'm going to try taking a quick look at the position thing too

**Comment by msullivan** [comment]
Oh, the missing spans are because postgres doesn't report the positions for that kind of invalid syntax error if it is a param and not a literal. Sigh.

**Comment by msullivan** [comment]
Great work on this.

File: edb/pgsql/resolver/expr.py [hunk 2]
```diff
@@ -591,6 +607,13 @@ def resolve_FuncCall(
         with_ordinality=call.with_ordinality,
     )
 
+    # HACK: for polymorphic functions that take params that have been
+    # extracted from string constants, we need to type-annotate the param.
+    if name[-1] == 'to_json' and len(res.args) >= 1:
+        res.args[0] = maybe_annotate_param(res.args[0], ctx=ctx)
```
[the remark this answers is not in the document]
Comment (aljazerzen) [reply edb/pgsql/resolver/expr.py]: Oh great, thanks for spotting.

File: edb/server/compiler/sql.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aljazerzen) [reply edb/server/compiler/sql.py]: Yes, that was left on my todo list.

**Pull request merged**